File: 算法竞赛进阶指南/0x00基本算法/AcWing113.py
# Forward declaration of compare API.
# def compare(a, b):
# @param a, b int
# @return bool
# return bool means whether a is less than b.

class Solution(object):
    def specialSort(self, N):
        """
        :type N: int
        :rtype: List[int]
        """
        a = []
        for i in range(N): a.append(i+1)
        for i in range(N):
            k = a[i]
            l, r = 0, k-1
            while l < r:
                mid = (l + r) >> 1
                if compare(k, a[mid]): r = mid
                else: l = mid + 1
            for j in range(i-1, l-1, -1): a[j+1] = a[j]
            a[l] = k
        return a

File: 0x00基本算法/0x04二分/AcWing113.cpp
// Forward declaration of compare API.
// bool compare(int a, int b);
// return bool means whether a is less than b.

class Solution {
public:
  vector<int> specialSort(int N) {
    vector<int> a;
    for (int i=0; i<N; i++) a.push_back(i+1);
    // 数学归纳法：假设k-1已经按要求排好，二分确定第k个元素放在哪一元素前面
    for (int i=0; i<N; i++) {
      int k = a[i]; 
      int l = 0, r = k-1;
      while (l < r) {
        int mid = (l + r) >> 1;
        if (compare(k, a[mid])) r = mid;
        else l = mid + 1;
      }
      // 找到k的放置位置l，需要先把[l,i-1]坐标上的元素往后挪一格
      for (int j=i-1; j>=l; j--) a[j+1] = a[j];
      a[l] = k;
    }
    return a;
  }
};